### ISSUE TITLE
E2E Tests

### ISSUE THREAD

**Issue by sidmohanty11** [RocketChat/EmbeddedChat#546]
We need to add e2e tests and have a mock api setup so that we can much better track any regressions or bugs

**Comment by JeffreytheCoder** [comment]
@sidmohanty11 Any tools in mind? I'd recommend Cypress for e2e tests and json-server for mock APIs

**Comment by sidmohanty11** [comment]
We already have a base setup using Playwright here (very similar to Cypress): https://github.com/RocketChat/EmbeddedChat/tree/develop/packages/e2e-react, and I was thinking if we can improve it from there.

**Comment by JeffreytheCoder** [comment]
There is one example test...we do need add more tests and some test coverage requirements

This would be a large epic - we can divide it into a task list like [this](https://github.com/RocketChat/Apps.Github22/issues/143)

**Comment by umangutkarsh** [comment]
I would like to work on this too. @sidmohanty11 @JeffreytheCoder maybe we can divide the work if that is okay.

**Comment by sidmohanty11** [comment]
Yes @JeffreytheCoder that would be awesome! We can also go API by API and their related feature set first and then target a bit harder features like realtime communications etc. What do you think?

**Comment by sidmohanty11** [comment]
@umangutkarsh that would be awesome! Lets divide it using a central issue and we can keep track from there only

**Comment by JeffreytheCoder** [comment]
> Yes @JeffreytheCoder that would be awesome! We can also go API by API and their related feature set first and then target a bit harder features like realtime communications etc. What do you think?

Sounds good! Could you create a central issue with a task list?

**Comment by JeffreytheCoder** [comment]
@sidmohanty11 Do we need mock APIs for e2e tests? My understanding is that e2e tests should simulate as close to real-world usage as possible. Although the RC server is an external service to EC, it's basically our entire backend.

Check out this [discussion](https://www.reddit.com/r/reactjs/comments/17y9vhy/e2e_testing_mock_api_or_use_real_endpoints/) on Reddit.

**Comment by umangutkarsh** [comment]
Can we use something like msw as mentioned in the discussion, only for api testing? With playwright to avoid hitting actual endpoints?

**Comment by Spiral-Memory** [comment]
Hey @sidmohanty11 , @JeffreytheCoder , @umangutkarsh ! 🌟 Since this issue is huge, and I personally want to learn writing e2e tests and learn more about automated workflows, I would love to contribute to this central issue if everyone is on board with it. 😊

**Comment by JeffreytheCoder** [comment]
Yes! @Spiral-Memory Let's do it together 🚀

**Comment by umangutkarsh** [comment]
I researched about mock service worker. Maybe we could use that? What do you all think?

**Comment by Spiral-Memory** [comment]
> Can we use something like msw with playwright as mentioned in the discussion, (only for api testing) to avoid hitting actual endpoints?

But why? I think it should be as close to real-world situations. Why do we want to do this? Any specific reason, @umangutkarsh ?

@JeffreytheCoder , what's your opinion?

**Comment by JeffreytheCoder** [comment]
Having mock APIs make our tests become integration tests, which also does the job since EC is just frontend. However, doing e2e tests means we don't have to create mock APIs and put EC into test by real user interaction scenarios. I would lean towards e2e tests

**Comment by umangutkarsh** [comment]
> > Can we use something like msw with playwright as mentioned in the discussion, (only for api testing) to avoid hitting actual endpoints?
> 
> But why? I think it should be as close to real-world situations. Why do we want to do this? Any specific reason, @umangutkarsh ? We already have avitechlab as hosted server on which we can hit for testing. Please correct me if i am missing something
> 
> If the concern is "server unavailable" or something like that, then we can test on real server, but when it is down or unavailable then we may have msw
> 
> @JeffreytheCoder , what's your opinion?

Avitechlab might not always be available I think. The certificate on the server might sometimes expire. And I think the tests should not be dependent on that.

**Comment by Spiral-Memory** [comment]
> > > Can we use something like msw with playwright as mentioned in the discussion, (only for api testing) to avoid hitting actual endpoints?
> > 
> > But why? I think it should be as close to real-world situations. Why do we want to do this? Any specific reason, @umangutkarsh ? We already have avitechlab as hosted server on which we can hit for testing. Please correct me if i am missing something
> > 
> > If the concern is "server unavailable" or something like that, then we can test on real server, but when it is down or unavailable then we may have msw
> > 
> > @JeffreytheCoder , what's your opinion?
> 
> Avitechlab might not always be available I think. The certificate on the server might sometimes expire. And I think the tests should not be dependent on that.

Yep, that's why i was thinking to have both approaches. For now, let's go with e2e test as suggested by @JeffreytheCoder 

Later we can handle the unavailable service issue by mocking.

**Comment by umangutkarsh** [comment]
So for testing only the user interaction scenarios, what do you think about using codegen which comes with playwright?

**Comment by umangutkarsh** [comment]
> So for testing only the user interaction scenarios, what do you think about using codegen which comes with playwright?

But for complex scenarios we might require manual coding for additional edge cases.

**Comment by Spiral-Memory** [comment]
> So for testing only the user interaction scenarios, what do you think about using codegen which comes with playwright?

Not sure, what exactly it does.

**Comment by umangutkarsh** [comment]
> > So for testing only the user interaction scenarios, what do you think about using codegen which comes with playwright?
> 
> Not sure, what exactly it does.

I mean it will generate the code for the user interactions which you can see in the Playwright Inspector window.

**Comment by Spiral-Memory** [comment]
Ohh @umangutkarsh then if required, it can be added..

**Comment by umangutkarsh** [comment]
```
import { test, expect } from '@playwright/test';

test('Send Message', async ({ page }) => {
  await page.goto('/');
  await page.getByPlaceholder('example@example.com').click();
  await page.getByPlaceholder('example@example.com').fill('tester123');
  await page.locator('input[type="password"]').click();
  await page.locator('input[type="password"]').fill('tester123');
  await page.getByRole('button', { name: 'Login' }).click();
  await page.getByPlaceholder('Message').click();
  await page.getByPlaceholder('Message').fill('send2');
  await page.locator('div:nth-child(3) > div:nth-child(2) > div > .ec-box > .ec-button').click();
});
```

```
import { test, expect } from '@playwright/test';

test('Login', async ({ page }) => {
  await page.goto('/');
  await page.getByRole('button', { name: 'JOIN' }).click();
  await page.getByPlaceholder('example@example.com').click();
  await page.getByPlaceholder('example@example.com').fill('tester123');
  await page.locator('input[type="password"]').click();
  await page.locator('input[type="password"]').fill('tester123');
  await page.locator('input[type="password"]').press('Enter');
});
```

a basic test of sending a message and login .

------
@sidmohanty11 @Spiral-Memory @JeffreytheCoder . some basic user interaction tests can be generated with codegen i think. But for testing I think we would require a shared account on the avitechlab server.
Any suggestions?

**Comment by Spiral-Memory** [comment]
@umangutkarsh once read our discussions in PR #570 regarding shared account. Also let's discuss more about this and work after we have a meeting and requirements have been clarified.

**Comment by umangutkarsh** [comment]
Oh, my bad. I didn't see this PR. 
Sure, @Spiral-Memory .

**Comment by devanshkansagra** [comment]
hey @Spiral-Memory and @sidmohanty11, I have seen @JeffreytheCoder 's pr, I have one doubt, do we have to test like the way @JeffreytheCoder did or do we have to test each and every function with of the EmbeddedChatApi(you mentioned above in the issue description) with the test inputs(or mock inputs)?

**Comment by devanshkansagra** [comment]
I am learning about mock testing now so this needs to be cleared to me

**Comment by Spiral-Memory** [comment]
Closing due to inactivity.
